mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,211 @@
+use std::{collections::HashMap, slice::Iter, sync::atomic::Ordering};
+
+use crate::{adapt::{Integer, MetricNameValueMap, PreparedMetric},
+            new_lib::{MetricType, MetricsJson}};
+use common::log::Tag;
+use fomat_macros::wite;
+use itertools::Itertools;
+use metrics::{Counter, Gauge, Key, KeyName, Label, Recorder, Unit};
+use metrics_core::ScopedString;
+use metrics_exporter_prometheus::formatting::key_to_parts;
+use metrics_util::registry::{GenerationalAtomicStorage, Registry};
+use std::fmt::Write;
+
+pub struct Snapshot {
+    pub counters: HashMap<String, HashMap<Vec<Label>, u64>>,
+    pub gauges: HashMap<String, HashMap<Vec<Label>, u64>>,
+    pub histograms: HashMap<String, HashMap<Vec<Label>, f64>>,
+}
+
+pub struct MmRecorder {
+    pub(crate) registry: Registry<Key, GenerationalAtomicStorage>,
+}
+
+impl Default for MmRecorder {
+    fn default() -> Self {
+        Self {
+            registry: Registry::new(metrics_util::registry::GenerationalStorage::atomic()),
+        }
+    }
+}
+
+impl MmRecorder {
+    pub fn get_metrics(&self) -> Snapshot {
+        let counters = self
+            .registry
+            .get_counter_handles()
+            .into_iter()
+            .map(|(key, counter)| {
+                counter.get_generation();
+                let value = counter.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let gauges = self
+            .registry
+            .get_gauge_handles()
+            .into_iter()
+            .map(|(key, gauge)| {
+                gauge.get_generation();
+                let value = gauge.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let histograms = self
+            .registry
+            .get_histogram_handles()
+            .into_iter()
+            .map(|(key, histogram)| {
+                histogram.get_generation();
+                let value: f64 = histogram.get_inner().data().iter().sum(); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    pub fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
```

Comment:
I think it's better to rename `prepare_tag_metrics` to `collect_tag_metrics` and `PreparedMetric` to `TagMetric`.